lib/iris/cube.py

```diff
@@ -790,6 +800,285 @@ def _is_single_item(testee):
 
 
+class CubeAttrsDict(MutableMapping):
+    """
+    A dict-like object for "Cube.attributes", which provides unified user access to
+    the combined cube 'local' and 'global' attributes, mimicking the behaviour of a
+    simple dict.
+
+    This supports all the regular methods of a 'dict'.
+    However, a few things such as the detailed print (repr) are different.
+
+    In addition, the 'locals' and 'globals' properties provide specific access to local
+    and global attributes, as regular
+    :class:`~iris.common.mixin.LimitedAttributeDict`\\s.
+
+    Notes
+    -----
+    For type testing, "issubclass(CubeAttrsDict, Mapping)" is True, but
+    "issubclass(CubeAttrsDict, dict)" is False.
+
+    Properties 'locals' and 'globals' are the two sets of cube attributes.  These are
+    both :class:`~iris.common.mixin.LimitedAttributeDict`.  The CubeAttrsDict object
+    contains *no* additional state of its own, but simply acts as a view on these two.
+
+    When reading (__getitem__, pop, popitem, keys, values etc), it contains all the
+    keys + values of both 'locals' and 'globals'.  When a key occurs in *both* 'locals'
+    and 'globals', the result is the local value.
+    See :meth:``~iris.cube.CubeAttrsDict__getitem__``.
+
+    When writing (__setitem__, setdefault, update, etc) to a key already present, the
+    existing entry in either 'locals' or 'globals' is updated.  If both are present,
+    'locals' is updated.  See :meth:``~iris.cube.CubeAttrsDict__setitem__``.
+
+    When writing to a new key, this generally updates 'locals'.  However, certain
+    specific names would never normally be 'data' attributes, and these are created as
+    'globals' instead.  These are listed in Appendix A of the
+    `CF Conventions: <https://cfconventions.org/>`_ .
+    At present, these are : 'conventions', 'featureType', 'history', 'title'.
+
+    Examples
+    --------
+
+        >>> from iris.cube import Cube
+        >>> cube = Cube([0])
+        >>> cube.attributes.update({"history": "from test-123", "mycode": 3})
+        >>> print(cube.attributes)
+        {'history': 'from test-123', 'mycode': 3}
+        >>> print(repr(cube.attributes))
+        CubeAttrsDict(globals={'history': 'from test-123'}, locals={'mycode': 3})
+
+        >>> cube.attributes['history'] += ' +added'
+        >>> print(repr(cube.attributes))
+        CubeAttrsDict(globals={'history': 'from test-123 +added'}, locals={'mycode': 3})
+
+        >>> cube.attributes.locals['history'] = 'per-variable'
+        >>> print(cube.attributes)
+        {'history': 'per-variable', 'mycode': 3}
+        >>> print(repr(cube.attributes))
+        CubeAttrsDict(globals={'history': 'from test-123 +added'}, locals={'mycode': 3, 'history': 'per-variable'})
+
+    """
+
+    def __init__(
+        self,
+        combined: Optional[Union[Mapping, str]] = "__unspecified",
+        locals: Optional[Mapping] = None,
+        globals: Optional[Mapping] = None,
+    ):
+        """
```

Comment:
OK I've worked out that I can add docstrings for `__setattr__` and `__getattr__`, and it will show them (which IIRC used to not work without special handling).
So I think I can simplify via that, and also make use of how the class-docs and init-docs look together to shorten + cross-link the info.
Watch this space ...